index.js

```diff
@@ -1,57 +1,79 @@
 'use strict'
 
-var level = require('level')
-var has = require('has')
-var pump = require('pump')
-var fs = require('fs')
-var net = require('net')
-var path = require('path')
-var multileveldown = require('multileveldown')
-
-module.exports = function (dir, opts) {
-  if (!opts) opts = {}
-  if (!has(opts, 'retry')) opts.retry = true
-
-  var sockPath = process.platform === 'win32'
+const level = require('level')
+const { pipeline: pump } = require('stream')
+const fs = require('fs')
+const net = require('net')
+const path = require('path')
+const multileveldown = require('multileveldown')
+
+module.exports = function (dir, opts = {}) {
+  const sockPath = process.platform === 'win32'
     ? '\\\\.\\pipe\\level-party\\' + path.resolve(dir)
     : path.join(dir, 'level-party.sock')
 
-  var client = multileveldown.client(opts)
+  opts = { retry: true, ...opts }
+
+  const client = multileveldown.client(opts)
 
   client.open(tryConnect)
 
   function tryConnect () {
-    if (!client.isOpen()) return
+    if (!client.isOpen()) {
+      return
+    }
 
-    var socket = net.connect(sockPath)
-    var connected = false
+    const socket = net.connect(sockPath)
+    let connected = false
 
     socket.on('connect', function () {
       connected = true
     })
 
-    // we pass socket as the ref option so we dont hang the event loop
+    // Pass socket as the ref option so we dont hang the event loop.
     pump(socket, client.createRpcStream({ ref: socket }), socket, function () {
-      if (!client.isOpen()) return
+      // TODO: err?
+
+      if (!client.isOpen()) {
+        return
+      }
 
-      var db = level(dir, opts, onopen)
+      const db = level(dir, opts, onopen)
 
       function onopen (err) {
         if (err) {
-          if (connected) return tryConnect()
-          return setTimeout(tryConnect, 100)
+          // TODO: This can cause an invisible retry loop that never completes
+          // and leads to memory leaks.
+          // TODO: What errors should be retried?
+          if (connected) {
+            tryConnect()
+          } else {
+            setTimeout(tryConnect, 100)
+          }
```

Comment:
This looks like this will constantly retry for all non masters taking quite a lot of cpu.

Reply:
Let's discuss in a separate issue or PR; if a fix is needed it would be nice to also add a test for it.

Reply:
Sure. Lets leave the TODO's though. I'd be happy to discuss all of them.